File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleTokenSession.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.Immutable
import kotlin.time.Duration
import kotlin.time.DurationUnit
import kotlin.time.toDuration

/**
 * A linked session: the token a run submits under, held as an absolute deadline because a counter
 * restarts at the wrong value after process death. Built only by
 * [UseSmileIDSampleTokenDecoder], so a session cannot exist without a token that decodes.
 */
@Immutable
data class UseSmileIDSampleTokenSession(
    /** A display handle — the token's `jti`, else a digest of it. Never a prefix of the credential. */
    val id: String,
    val token: String,
    val issuedAtMillis: Long,
    val expiresAtMillis: Long,
    val bindings: UseSmileIDSampleTokenBindings,
    /**
     * The partner the token was minted for, from its own `partner_id` claim. The authority for a
     * submission's identity: sending a locally configured id alongside a real token is how a signed
     * request gets a 401. Never logged — a partner id is on this repo's never-commit list.
     */
    val partnerId: String? = null,
) {

    fun remaining(nowMillis: Long): Duration = (expiresAtMillis - nowMillis).coerceAtLeast(0L).milliseconds

    fun hasExpired(nowMillis: Long): Boolean = nowMillis >= expiresAtMillis

    /**
     * 1f on a fresh session down to 0f on an expired one, over the token's own span, for the nav bar's
     * ring. The decoder rejects a token whose `exp` is not after its `iat`, so the span is positive by
     * construction — but this constructor is public, and a zero span would divide to `NaN`, which
     * `coerceIn` passes straight through to the ring.
     */
    fun progress(nowMillis: Long): Float {
        val span = (expiresAtMillis - issuedAtMillis).coerceAtLeast(1L)
        return (remaining(nowMillis).inWholeMilliseconds.toFloat() / span).coerceIn(0f, 1f)
    }

    /** Redacted: the generated `toString` is how a bearer credential reaches a log or a crash report. */
    override fun toString(): String = "UseSmileIDSampleTokenSession(id=$id, expiresAtMillis=$expiresAtMillis)"
}

/** `m:ss`, growing an hours part when the span needs one — an 8h token reads 7:59:12, not 479:12. */
fun Duration.toCountdown(): String {
    val total = inWholeSeconds
    val minutes = (total % SECONDS_PER_HOUR) / SECONDS_PER_MINUTE
    val seconds = total % SECONDS_PER_MINUTE
    val hours = total / SECONDS_PER_HOUR
    return if (hours > 0) {
        "$hours:${minutes.padded()}:${seconds.padded()}"
    } else {
        "$minutes:${seconds.padded()}"
    }
}

private fun Long.padded(): String = toString().padStart(2, '0')

private const val SECONDS_PER_MINUTE = 60
private const val SECONDS_PER_HOUR = 3600

/** Exact rather than divided through a Double, so the countdown and the ring agree to the millisecond. */
private val Long.milliseconds: Duration get() = toDuration(DurationUnit.MILLISECONDS)

File: flutter/sample_ui/lib/src/state/use_smileid_sample_token_session.dart
import 'package:flutter/foundation.dart';

import '../model/use_smileid_sample_environment.dart';
import 'use_smileid_sample_token_decoder.dart';

/// A linked session, held as an absolute deadline; built by the decoder, so it cannot exist without a token that decodes.
@immutable
class UseSmileIDSampleTokenSession {
  /// Public so a test can build one; the app only ever gets one from [UseSmileIDSampleTokenDecoder].
  const UseSmileIDSampleTokenSession({
    required this.id,
    required this.token,
    required this.issuedAtMillis,
    required this.expiresAtMillis,
    required this.bindings,
    required this.environment,
    this.partnerId,
  });

  /// A display handle: the token's `jti`, else a digest of it. Never a prefix of the credential.
  final String id;

  /// The bearer credential a run submits under; never logged, rendered or tagged.
  final String token;

  /// The token's `iat`, in epoch milliseconds.
  final int issuedAtMillis;

  /// The token's `exp`, in epoch milliseconds.
  final int expiresAtMillis;

  /// What the token binds.
  final UseSmileIDSampleTokenBindings bindings;

  /// The partner the token was minted for, which wins over the local profile; never logged.
  final String? partnerId;

  /// From the token's own `api_url` claim; the decoder refuses a token it cannot place.
  final UseSmileIDSampleEnvironment environment;

  /// What is left at [nowMillis], never negative.
  Duration remaining(int nowMillis) => Duration(
    milliseconds: expiresAtMillis - nowMillis < 0
        ? 0
        : expiresAtMillis - nowMillis,
  );

  /// True from the deadline on.
  bool hasExpired(int nowMillis) => nowMillis >= expiresAtMillis;

  /// 1 fresh to 0 expired over the token's own span, for the nav bar's ring; a zero span cannot divide to NaN.
  double progress(int nowMillis) {
    final int span = expiresAtMillis - issuedAtMillis < 1
        ? 1
        : expiresAtMillis - issuedAtMillis;
    return (remaining(nowMillis).inMilliseconds / span).clamp(0, 1).toDouble();
  }

  @override
  bool operator ==(Object other) =>
      other is UseSmileIDSampleTokenSession && other.token == token;

  @override
  int get hashCode => token.hashCode;

  /// Redacted: a generated description is how a bearer credential reaches a log.
  @override
  String toString() =>
      'UseSmileIDSampleTokenSession(id: $id, environment: ${environment.label}, '
      'expiresAtMillis: $expiresAtMillis)';
}

/// `m:ss`, growing an hours part when the span needs one: an 8h token reads 7:59:12, not 479:12.
String useSmileIDSampleCountdown(Duration remaining) {
  final int total = remaining.inSeconds;
  final int hours = total ~/ Duration.secondsPerHour;
  final int minutes =
      (total % Duration.secondsPerHour) ~/ Duration.secondsPerMinute;
  final String seconds = (total % Duration.secondsPerMinute).toString().padLeft(
    2,
    '0',
  );
  return hours > 0
      ? '$hours:${minutes.toString().padLeft(2, '0')}:$seconds'
      : '$minutes:$seconds';
}
